pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -920,6 +956,33 @@ func (m *ManagerImpl) callPreStartContainerIfNeeded(podUID, contName, resource s
 	return nil
 }
 
+// callGetPreferredAllocationIfAvailable issues GetPreferredAllocation grpc
+// call for device plugin resource with GetPreferredAllocationAvailable option set.
+func (m *ManagerImpl) callGetPreferredAllocationIfAvailable(podUID, contName, resource string, available, mustInclude sets.String, size int) (sets.String, error) {
+	eI, ok := m.endpoints[resource]
+	if !ok {
+		return nil, fmt.Errorf("endpoint not found in cache for a registered resource: %s", resource)
+	}
+
+	if eI.opts == nil || !eI.opts.GetPreferredAllocationAvailable {
+		klog.V(4).Infof("Plugin options indicate to skip GetPreferredAllocation for resource: %s", resource)
+		return nil, nil
+	}
+
+	devices := m.podDevices.containerDevices(podUID, contName, resource)
+	if devices == nil {
+		return nil, fmt.Errorf("no devices found allocated in local cache for pod %s, container %s, resource %s", podUID, contName, resource)
+	}
```

Comment:
Could you explain why this check and returning an error is necessary here? It's not immediately clear to me, thanks.

Reply:
I didn't actually look too closely at this when adding it.

It is copy-pasted from the `callPreStartContainerIfNeeded()` function above, which is similar in nature to this one (i.e. only call out to the plugin it if the options set on the plugin indicate to).

Let me take a closer look at everything in this function to see if we need it.

Reply:
Just looked. Good catch. Not only is it unnecessary, it would have caused problems once I started writing tests that actually exercise this function (because it's highly likely that no devices will be allocated at the time this is called!).

Reply:
Fixed in latest